pkgs/development/tools/profiling/austin/default.nix

```diff
@@ -0,0 +1,71 @@
+{ stdenv, lib, fetchFromGitHub, fetchpatch, testers
+, autoreconfHook, libbfd, libiberty, libunwind, zlib
+, austin  # For testVersion
+, python3Packages  # For austin-python
+}:
+
+stdenv.mkDerivation rec {
+  pname = "austin";
+  version = "3.4.1";
+
+  src = fetchFromGitHub {
+    owner = "p403n1x87";
+    repo = "austin";
+    rev = "v${version}";
+    hash = "sha256-R/LxGHapPjaaTbJh6/CxT49WvX3zg7u+mATQGZCEXtk=";
+  };
+
+  patches = [
+    # https://github.com/P403n1x87/austin/issues/152
+    (fetchpatch {
+      url = "https://github.com/P403n1x87/austin/commit/1e26986ecd8981460af5b9bd6526d6560f9c16a0.patch";
+      hash = "sha256-c+r1iLMXkPU3Mzn8qyZv0EDnTbrxZQrWGdRkebNTN/4=";
+    })
+  ];
+
+  postPatch = ''
+    # Allow linking with dynamic libraries. Replace "-l:libfoo.a" -> "-lfoo".
+    sed -ri 's/-l:lib([^.]*).a/-l\1/g' configure.ac
+  '';
+
+  nativeBuildInputs = [ autoreconfHook ];
+
+  buildInputs = [
+    libbfd
+    libiberty
+    libunwind
+    zlib
+  ];
+
+  checkInputs = with python3Packages; [
+    (austin-python.override { austin = null; })
+    flaky
+    pycparser
+    pytest
+  ];
+
+  doCheck = true;
+
+  checkPhase = ''
+    NIX_HARDENING_ENABLE="" NIX_CFLAGS_COMPILE="$NIX_CFLAGS_COMPILE -fPIC" \
+      pytest -k "not test_cli_permissions"
```

Comment:
Please adjust this to pytestCheckHook

Reply:
How do you provide the requisite environment variables to pytest then? I've read the hook's source code and it doesn't seem to support this. I could `export` in preCheckPhase but then the environment leaks unless I also restore in postCheckPhase, and that seems like a lot of effort for very marginal wins.

Reply:
> I could `export` in preCheckPhase

```
preCheck = ''
  export NIX_HARDENING_ENABLE=""
  export NIX_CFLAGS_COMPILE="$NIX_CFLAGS_COMPILE -fPIC"
'';
```

> then the environment leaks to everything running later in the derivation

That is not a problem. I don't think there is running anything after checkPhase. So tests do not modify $out.

Reply:
How do you know that nothing runs in postCheck? How do you know nothing will run in postCheck in the future? This is more fragile for roughly 0 benefits because in this instance it's trivial to run pytest manually.

Happy to migrate this to pytestCheckHook once it supports the features that it would need to support to make using it here not fragile. Right now it would be leaking environment to an unknown amount of other hooks. FWIW installPhase and fixupPhase both run after checkPhase.

Reply:
> How do you know that nothing runs in postCheck?

I look at the source code of the hook and that it does not add anything there https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/interpreters/python/hooks/pytest-check-hook.sh#L59-L74

> How do you know nothing will run in postCheck in the future?

You don't but if we ever add something there it will be carefully be considered. It is unlikely because it wouldn't run when checkPhase is overwritten. I could imagine adding some cleanup for pytestCheckHook there in the future if we ever need it. Nothing you need to worry about.

> This is more fragile for roughly 0 benefits because in this instance it's trivial to run pytest manually.

It is not fragile at all. A lot of packages use it and it is proven to make python maintenance easier by not modifying pytest invocations in packages.

> Right now it would be leaking environment to an unknown amount of other hooks.

We are not compiling anything after checkPhase so that is not a problem.

> FWIW installPhase and fixupPhase both run after checkPhase.

that is true for mkDerivation but [mkPythonPackages maps checkPhase to installCheckPhase](https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/interpreters/python/mk-python-derivation.nix#L197) so [only distPhase is run afterwards](https://github.com/NixOS/nixpkgs/blob/192e7bbb9240d8ceac534dd3b6ead7aa175dc3ce/doc/stdenv/stdenv.chapter.md#variables-affecting-phase-control-variables-affecting-phase-control)